### PR TITLE
Create a FLOSS build for FDroid

### PR DESCRIPTION
**Pull request by maxammann**
* Added a way to sign the build using the local gradle properties
* Add a floss build to CI
* Upgraded geolocator (I fixed a bug upstream and also removed a usage of GMS)

This is not enough to release to fdroid right now. Store descriptions etc are missing. That will be done in a separate PR.

#380

### CODE REVIEW

File: frontend/android/app/build.gradle [hunk 1]
```diff
@@ -37,20 +50,40 @@ android {
     }
 
     defaultConfig {
-        applicationId "de.nrw.it.giz.ehrensache.bayern.android"
+        if (dartEnvironmentVariables.ANDROID_FLOSS) {
+            applicationId "app.ehrenamtskarte.floss"
```
Comment (Schedulaar) [inline frontend/android/app/build.gradle:54]:
```suggestion
            applicationId "app.ehrenamtskarte.bayern.floss"
```

File: frontend/android/app/build.gradle [hunk 2]
```diff
@@ -61,10 +94,24 @@ android {
         }
     }
 
-    // Exclude mapbox-android-accounts because of its proprietary license
     applicationVariants.all { variant ->
+        // Exclude mapbox-android-accounts because of its proprietary license
         variant.getCompileConfiguration().exclude group: 'com.mapbox.mapboxsdk', module: 'mapbox-android-accounts'
         variant.getRuntimeConfiguration().exclude group: 'com.mapbox.mapboxsdk', module: 'mapbox-android-accounts'
```
Comment (Schedulaar) [inline frontend/android/app/build.gradle:100]: Is this still necessary?

**Review by Schedulaar — APPROVED** [review]

File: frontend/android/app/build.gradle (hunk 2 above)
Comment (maxammann) [reply frontend/android/app/build.gradle]: I removed it :)

File: frontend/android/app/build.gradle [hunk 3]
```diff
@@ -61,10 +92,20 @@ android {
         }
     }
 
-    // Exclude mapbox-android-accounts because of its proprietary license
     applicationVariants.all { variant ->
-        variant.getCompileConfiguration().exclude group: 'com.mapbox.mapboxsdk', module: 'mapbox-android-accounts'
-        variant.getRuntimeConfiguration().exclude group: 'com.mapbox.mapboxsdk', module: 'mapbox-android-accounts'
+        // If we create a FLOSS build, then we want to exclude GMS
+        if (dartEnvironmentVariables.ANDROID_FLOSS) {
```
Comment (maxammann) [inline frontend/android/app/build.gradle:97]: This is one way to configure builds from the `flutter` cli tool.  Not sure yet how we want to do this.
There are multiple dimensions for creating builds:

* Dimension 1: Release, Debug, Profile build
* Dimension 2: Production, Staging, Local Server
* Dimension 3: Nürnberg, Bavaria, BaWü etc.
* Dimension 4: FLOSS or GAPPS
Any combination of the 4 dimension makes sense:

* Release, Production, Nürnberg, FLOSS
* Debug, Staging, Nürnberg, GAPPS

So we need to come up with a good concept.

File: frontend/android/app/build.gradle [hunk 4]
```diff
@@ -61,10 +92,20 @@ android {
...
+            variant.getCompileConfiguration().exclude group: 'com.google.android.gms', module: 'play-services-location'
+            variant.getRuntimeConfiguration().exclude group: 'com.google.android.gms', module: 'play-services-location'
+        }
+    }
+    
+    packagingOptions {
+        
+        // Exclude x86 because the app size in F-Droid is limited to 30MB
+        if (dartEnvironmentVariables.SHRINK_NATIVE) {
```
Comment (maxammann) [inline frontend/android/app/build.gradle:106]: A shrinked build is not possible to execute on an x86 emulator

File: frontend/lib/environment.dart [hunk 5]
```diff
@@ -1,3 +1,3 @@
 class EnvironmentConfig {
-  static const ANDROID_FLOSS = bool.fromEnvironment('ANDROID_FLOSS', defaultValue: false);
+  static const androidFloss = bool.fromEnvironment('ANDROID_FLOSS', defaultValue: false);
```
Comment (Schedulaar) [inline frontend/lib/environment.dart:2]: quite nice :D

File: frontend/android/app/build.gradle (hunk 3 above)
Comment (Schedulaar) [reply frontend/android/app/build.gradle:97]: Yep, but it seems the whitelabeling stuff should be discussed in another issue.
For now, the approach with a env variable seems very fine to me.

**Review by Schedulaar — APPROVED** [review]

File: .circleci/config.yml [hunk 6]
```diff
@@ -140,6 +143,35 @@ jobs:
       - store_artifacts:
           path: build/app/outputs/flutter-apk/
 
+  frontend-android-floss-build:
```
Comment (klinzo) [inline .circleci/config.yml:146]: Can't we just make the `frontend-android-build` more generic such that we pass the option `FLOSS` and don't have to define it twice? It is mostly duplicated

File: frontend/lib/location/location_ffi.dart [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import 'package:flutter/services.dart';
+
+const platform = MethodChannel('app.ehrenamtskarte/location');
+
+/// Checks whether location services are enabled
+/// without using Google Play Services
+Future<bool> isNonGoogleLocationServiceEnabled() async {
```
Comment (klinzo) [inline frontend/lib/location/location_ffi.dart:7]: Does this fix #432 ?

File: frontend/lib/location/location_ffi.dart [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import 'package:flutter/services.dart';
```
Comment (klinzo) [inline frontend/lib/location/location_ffi.dart:1]: What does `location_ffi` stand for?

File: frontend/android/app/src/main/kotlin/app/ehrenamtskarte/MainActivity.kt [hunk 9]
```diff
@@ -1,13 +1,32 @@
...
+    private val CHANNEL = "app.ehrenamtskarte/location"
+
+    override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
+        super.configureFlutterEngine(flutterEngine)
+        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL).setMethodCallHandler {
+                call, result ->
+            if (call.method == "isLocationServiceEnabled") {
+                val locationManager = context.getSystemService(LOCATION_SERVICE) as LocationManager
+                val gpsEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
+                val networkEnabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)
```
Comment (klinzo) [inline frontend/android/app/src/main/kotlin/app/ehrenamtskarte/MainActivity.kt:23]: Why do we have to check for network here?

**Review by klinzo — APPROVED** [review]

File: frontend/lib/location/location_ffi.dart (hunk 7 above)
Comment (maxammann) [reply frontend/lib/location/location_ffi.dart:7]: I dont think so

File: frontend/android/app/src/main/kotlin/app/ehrenamtskarte/MainActivity.kt (hunk 9 above)
Comment (maxammann) [reply frontend/android/app/src/main/kotlin/app/ehrenamtskarte/MainActivity.kt:23]: This was copy pasted from some stackoverflow. Ill review it again though.

Comment (maxammann) [reply frontend/android/app/src/main/kotlin/app/ehrenamtskarte/MainActivity.kt:23]: This is not  a network check, but checks whether the network provider is enabled: https://developer.android.com/reference/android/location/LocationManager

File: frontend/lib/location/location_ffi.dart (hunk 8 above)
Comment (maxammann) [reply frontend/lib/location/location_ffi.dart:1]: ffi stands for foreign function interface.

**Pull request merged**